<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>noisyR count matrix approach workflow</title>


<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(title);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">noisyR count matrix approach workflow</h1>



<div style="text-align: justify">
<p>The <em>noisyR</em> package is an end-to-end pipeline for quantifying and removing technical noise from HTS datasets. The three main pipeline steps are:</p>
<ol style="list-style-type: decimal">
<li>similarity calculation<br />
</li>
<li>noise quantification<br />
</li>
<li>noise removal</li>
</ol>
<p>Each step can be finely tuned using hyperparameters; optimal, data-driven values for these parameters are also determined.</p>
<p>The package and some applications are described in more detail in <a href="https://www.biorxiv.org/content/10.1101/2021.01.17.427026v2">this paper</a> and is actively maintained on <a href="https://github.com/Core-Bioinformatics/noisyR" class="uri">https://github.com/Core-Bioinformatics/noisyR</a>.</p>
<p>The <strong>count matrix approach</strong> uses the original, un-normalised count matrix, as provided after alignment and feature quantification; each sample is processed individually, only the relative expressions across samples are compared. Relying on the hypothesis that the majority of genes are not DE, most of the evaluations are expected to point towards a high similarity across samples.</p>
<h1 id="installation">Installation</h1>
<p>To install the package, first install all bioconductor dependencies:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" title="1">packages.bioc &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;preprocessCore&quot;</span>,</a>
<a class="sourceLine" id="cb1-2" title="2">                   <span class="st">&quot;IRanges&quot;</span>,</a>
<a class="sourceLine" id="cb1-3" title="3">                   <span class="st">&quot;GenomicRanges&quot;</span>,</a>
<a class="sourceLine" id="cb1-4" title="4">                   <span class="st">&quot;Rsamtools&quot;</span>)</a>
<a class="sourceLine" id="cb1-5" title="5">new.packages.bioc &lt;-<span class="st"> </span>packages.bioc[<span class="op">!</span>(packages.bioc <span class="op">%in%</span><span class="st"> </span><span class="kw">installed.packages</span>()[,<span class="st">&quot;Package&quot;</span>])]</a>
<a class="sourceLine" id="cb1-6" title="6"><span class="cf">if</span>(<span class="kw">length</span>(new.packages.bioc)){</a>
<a class="sourceLine" id="cb1-7" title="7">  <span class="cf">if</span> (<span class="op">!</span><span class="kw">requireNamespace</span>(<span class="st">&quot;BiocManager&quot;</span>, <span class="dt">quietly =</span> <span class="ot">TRUE</span>))</a>
<a class="sourceLine" id="cb1-8" title="8">    <span class="kw">install.packages</span>(<span class="st">&quot;BiocManager&quot;</span>)</a>
<a class="sourceLine" id="cb1-9" title="9">  BiocManager<span class="op">::</span><span class="kw">install</span>(new.packages.bioc)</a>
<a class="sourceLine" id="cb1-10" title="10">}</a></code></pre></div>
<p>Then, you can install <em>noisyR</em> (and all its other dependencies) from CRAN:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" title="1"><span class="kw">install.packages</span>(<span class="st">&quot;noisyr&quot;</span>)</a></code></pre></div>
<p>To install the latest stable version from GitHub, first install CRAN dependencies:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" title="1">packages.cran &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;utils&quot;</span>,</a>
<a class="sourceLine" id="cb3-2" title="2">                   <span class="st">&quot;grDevices&quot;</span>,</a>
<a class="sourceLine" id="cb3-3" title="3">                   <span class="st">&quot;tibble&quot;</span>,</a>
<a class="sourceLine" id="cb3-4" title="4">                   <span class="st">&quot;dplyr&quot;</span>,</a>
<a class="sourceLine" id="cb3-5" title="5">                   <span class="st">&quot;magrittr&quot;</span>,</a>
<a class="sourceLine" id="cb3-6" title="6">                   <span class="st">&quot;ggplot2&quot;</span>,</a>
<a class="sourceLine" id="cb3-7" title="7">                   <span class="st">&quot;philentropy&quot;</span>,</a>
<a class="sourceLine" id="cb3-8" title="8">                   <span class="st">&quot;doParallel&quot;</span>,</a>
<a class="sourceLine" id="cb3-9" title="9">                   <span class="st">&quot;foreach&quot;</span>)</a>
<a class="sourceLine" id="cb3-10" title="10">new.packages.cran &lt;-<span class="st"> </span>packages.cran[<span class="op">!</span>(packages.cran <span class="op">%in%</span><span class="st"> </span><span class="kw">installed.packages</span>()[,<span class="st">&quot;Package&quot;</span>])]</a>
<a class="sourceLine" id="cb3-11" title="11"><span class="cf">if</span>(<span class="kw">length</span>(new.packages.cran))</a>
<a class="sourceLine" id="cb3-12" title="12">  <span class="kw">install.packages</span>(new.packages.cran)</a>
<a class="sourceLine" id="cb3-13" title="13"></a>
<a class="sourceLine" id="cb3-14" title="14"><span class="cf">if</span> (<span class="op">!</span><span class="kw">requireNamespace</span>(<span class="st">&quot;devtools&quot;</span>, <span class="dt">quietly =</span> <span class="ot">TRUE</span>))</a>
<a class="sourceLine" id="cb3-15" title="15">  <span class="kw">install.packages</span>(<span class="st">&quot;devtools&quot;</span>)</a>
<a class="sourceLine" id="cb3-16" title="16">devtools<span class="op">::</span><span class="kw">install_github</span>(<span class="st">&quot;Core-Bioinformatics/noisyR&quot;</span>)</a></code></pre></div>
<h1 id="preprocessing">Preprocessing</h1>
<p>First, load <em>noisyR</em>:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" title="1"><span class="kw">library</span>(noisyr)</a></code></pre></div>
<p>For this demonstration we will be using a subset of the count matrix for an experiment included in <a href="https://www.sciencedirect.com/science/article/pii/S2405471219301152">a 2019 paper by Yang et al</a>. Rows represent genes/features and columns represent samples:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" title="1">counts.in &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;counts_raw.csv&quot;</span>, <span class="dt">package =</span> <span class="st">&quot;noisyr&quot;</span>)</a>
<a class="sourceLine" id="cb5-2" title="2">df &lt;-<span class="st"> </span><span class="kw">read.csv</span>(counts.in, <span class="dt">row.names =</span> <span class="dv">1</span>)</a>
<a class="sourceLine" id="cb5-3" title="3"><span class="kw">str</span>(df)</a>
<a class="sourceLine" id="cb5-4" title="4"><span class="kw">head</span>(df)</a></code></pre></div>
<p>Note that when reading from a file R typically returns a data frame. To convert to a matrix we use the function <em>cast_matrix_to_numeric()</em>. This also converts values to numeric (in case they were read as characters). Any values that are not coercible to numeric are replaced by 0.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" title="1">expression.matrix &lt;-<span class="st"> </span>noisyr<span class="op">::</span><span class="kw">cast_matrix_to_numeric</span>(df)</a></code></pre></div>
<h1 id="running-noisyr">Running <em>noisyR</em></h1>
<p>The full <em>noisyR</em> pipeline can be run through <em>noisyr()</em>, choosing “counts” for the count matrix approach (internally calls <em>noisyr_counts()</em>). The user can pass many arguments to this function, which alter the behaviour of the different pipeline steps, as discussed in the breakdown below.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" title="1">expression.matrix.denoised.standard &lt;-<span class="st"> </span>noisyr<span class="op">::</span><span class="kw">noisyr</span>(</a>
<a class="sourceLine" id="cb7-2" title="2">  <span class="dt">approach_for_similarity_calculation =</span> <span class="st">&quot;counts&quot;</span>, </a>
<a class="sourceLine" id="cb7-3" title="3">  <span class="dt">expression.matrix =</span> df</a>
<a class="sourceLine" id="cb7-4" title="4">)</a></code></pre></div>
<p>The output of the noise removal is a denoised matrix that can be passed on to other methods for downstream analysis.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" title="1"><span class="kw">head</span>(expression.matrix.denoised.standard)</a>
<a class="sourceLine" id="cb8-2" title="2"><span class="kw">apply</span>(expression.matrix.denoised.standard, <span class="dv">2</span>, min)</a></code></pre></div>
<h1 id="pipeline-breakdown">Pipeline breakdown</h1>
<p>While for most applications, running <em>noisyr()</em> is sufficient, it may be useful to run individual pipeline steps manually. For example, the user may want to create summary figures for the denoising, store the noise thresholds obtained or other intermediary outputs, or try out different options without rerunning all steps. In this section, we look deeper into <em>noisyr()</em> and break down the three main steps that are performed.</p>
<h2 id="similarity-calculation">Similarity calculation</h2>
<p>We can then run the similarity calculation using <em>calculate_expression_similarity_counts()</em>:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" title="1">expression.summary &lt;-<span class="st"> </span>noisyr<span class="op">::</span><span class="kw">calculate_expression_similarity_counts</span>(</a>
<a class="sourceLine" id="cb9-2" title="2">  <span class="dt">expression.matrix =</span> expression.matrix, </a>
<a class="sourceLine" id="cb9-3" title="3">  <span class="dt">similarity.measure =</span> <span class="st">&quot;correlation_pearson&quot;</span></a>
<a class="sourceLine" id="cb9-4" title="4">)</a>
<a class="sourceLine" id="cb9-5" title="5"><span class="kw">str</span>(expression.summary)</a></code></pre></div>
<p>Users can select a similarity measure to assess the localised consistency in expression across samples (dissimilarity measures are inverted). See the <em>philentropy</em> package documentation for more information on the different distances. The full list of available metrics can be viewed by:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" title="1">noisyr<span class="op">::</span><span class="kw">get_methods_correlation_distance</span>()</a></code></pre></div>
<p>By default, the window length is 10% of the number of rows in the matrix, as it has proven effective empirically. A different window can be specified by the <em>n.elements.per.window</em> parameter. The optimal window length can also be estimated by seeking stability of output (but this can be computationally intensive for large datasets):</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" title="1">noisyr<span class="op">::</span><span class="kw">optimise_window_length</span>(</a>
<a class="sourceLine" id="cb11-2" title="2">  <span class="dt">expression.matrix =</span> expression.matrix,</a>
<a class="sourceLine" id="cb11-3" title="3">  <span class="dt">similarity.measure =</span> <span class="st">&quot;correlation_pearson&quot;</span></a>
<a class="sourceLine" id="cb11-4" title="4">)</a></code></pre></div>
<p>Window length optimisation can be turned on using the <em>optimise.window.length.logical</em> parameter in <em>noisyr()</em> or <em>noisyr_counts()</em>.</p>
<p>Plots of the abundance-correlation relation can be generated through the <em>plot_expression_similarity()</em> function:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" title="1">plotlist &lt;-<span class="st"> </span>noisyr<span class="op">::</span><span class="kw">plot_expression_similarity</span>(</a>
<a class="sourceLine" id="cb12-2" title="2">  <span class="dt">expression.summary =</span> expression.summary)</a>
<a class="sourceLine" id="cb12-3" title="3">plotlist[[<span class="dv">1</span>]]</a></code></pre></div>
<blockquote>
<p>As expected, we observe low correlation values for low abundances and a steady increase towards 1 as the abundance increases. This is based on the expectation that most genes are not differentially expressed and have consistent expression, but at low abundances the stochastic nature of transcription and sequencing gives rise to noise. The local maximum at very low abundances is due to strings of zeros driving the correlation higher than expected.</p>
</blockquote>
<p>These are ggplot objects, and can thus be modified and combined intuitively. For example, plotting all the line plots together:</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" title="1">plotdf.line &lt;-<span class="st"> </span>tibble<span class="op">::</span><span class="kw">tibble</span>()</a>
<a class="sourceLine" id="cb13-2" title="2"><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="dv">4</span>){</a>
<a class="sourceLine" id="cb13-3" title="3">  lineid &lt;-<span class="st"> </span>i <span class="op">*</span><span class="st"> </span><span class="dv">2</span> <span class="op">-</span><span class="st"> </span><span class="dv">1</span></a>
<a class="sourceLine" id="cb13-4" title="4">  plotdf.line &lt;-<span class="st"> </span><span class="kw">rbind</span>(</a>
<a class="sourceLine" id="cb13-5" title="5">    plotdf.line, </a>
<a class="sourceLine" id="cb13-6" title="6">    dplyr<span class="op">::</span><span class="kw">mutate</span>(plotlist[[lineid]]<span class="op">$</span>data,</a>
<a class="sourceLine" id="cb13-7" title="7">                  <span class="dt">Sample=</span><span class="kw">colnames</span>(expression.matrix)[i]))</a>
<a class="sourceLine" id="cb13-8" title="8">}</a>
<a class="sourceLine" id="cb13-9" title="9"></a>
<a class="sourceLine" id="cb13-10" title="10">ggplot2<span class="op">::</span><span class="kw">ggplot</span>(plotdf.line) <span class="op">+</span></a>
<a class="sourceLine" id="cb13-11" title="11"><span class="st">    </span>ggplot2<span class="op">::</span><span class="kw">theme_minimal</span>() <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb13-12" title="12"><span class="st">    </span>ggplot2<span class="op">::</span><span class="kw">geom_line</span>(ggplot2<span class="op">::</span><span class="kw">aes</span>(<span class="dt">x=</span>x, <span class="dt">y=</span>y, <span class="dt">colour=</span>Sample)) <span class="op">+</span></a>
<a class="sourceLine" id="cb13-13" title="13"><span class="st">    </span>ggplot2<span class="op">::</span><span class="kw">geom_smooth</span>(ggplot2<span class="op">::</span><span class="kw">aes</span>(x,y,<span class="dt">colour=</span>Sample), <span class="dt">method=</span><span class="st">&quot;loess&quot;</span>,</a>
<a class="sourceLine" id="cb13-14" title="14">                         <span class="dt">formula=</span> y <span class="op">~</span><span class="st"> </span>x, <span class="dt">span=</span><span class="fl">0.1</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb13-15" title="15"><span class="st">    </span>ggplot2<span class="op">::</span><span class="kw">ylim</span>(<span class="dv">0</span><span class="op">:</span><span class="dv">1</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb13-16" title="16"><span class="st">    </span>ggplot2<span class="op">::</span><span class="kw">xlab</span>(<span class="st">&quot;log2(expression)&quot;</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb13-17" title="17"><span class="st">    </span>ggplot2<span class="op">::</span><span class="kw">ylab</span>(<span class="st">&quot;Pearson correlation&quot;</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb13-18" title="18"><span class="st">    </span>ggplot2<span class="op">::</span><span class="kw">geom_hline</span>(<span class="dt">yintercept=</span><span class="fl">0.25</span>, <span class="dt">color=</span><span class="st">&quot;black&quot;</span>)</a></code></pre></div>
<h2 id="noise-quantification">Noise quantification</h2>
<p>Using the output of the similarity calculation, we can compute the signal to noise threshold in each sample:</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" title="1">noise.thresholds &lt;-<span class="st"> </span>noisyr<span class="op">::</span><span class="kw">calculate_noise_threshold_base</span>(<span class="dt">expression =</span> expression.summary)</a>
<a class="sourceLine" id="cb14-2" title="2">noise.thresholds</a></code></pre></div>
<p>Here we used the default parameters: a similarity threshold of 0.25 and the Boxplot-IQR method. There are several methods available, which can be viewed with the <em>get_methods_calculate_noise_threshold()</em> function:</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" title="1">noisyr<span class="op">::</span><span class="kw">get_methods_calculate_noise_threshold</span>()</a></code></pre></div>
<p>The first three methods are just calculating the minimum of the density plot for all genes (a common, fast approach). This usually provides a rough, overestimated signal to noise threshold.</p>
<p>The rest of the methods use either the (smoothed) line plot or the boxplot to find the noise threshold given a similarity (correlation/distance) threshold.</p>
<p>It is recommended that the method with the least coefficient of variation across all samples is chosen for noise removal. This can also be applied to compute the correlation/distance threshold instead of supplying it manually, which is especially useful for non-correlation measures which don’t have a standard range.</p>
<p>For example, by looking to minimise the coefficient of variation, we get a correlation threshold of 0.21 and the loess10 smoothing method for this dataset:</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" title="1">similarity.threshold.sequence &lt;-<span class="st"> </span><span class="kw">seq</span>(<span class="fl">0.2</span>, <span class="fl">0.3</span>, <span class="dt">by=</span><span class="fl">0.01</span>)</a>
<a class="sourceLine" id="cb16-2" title="2">stats.table &lt;-<span class="st"> </span>noisyr<span class="op">::</span><span class="kw">calculate_noise_threshold_method_statistics</span>(</a>
<a class="sourceLine" id="cb16-3" title="3">  <span class="dt">expression =</span> expression.summary,</a>
<a class="sourceLine" id="cb16-4" title="4">  <span class="dt">similarity.threshold.sequence =</span> similarity.threshold.sequence</a>
<a class="sourceLine" id="cb16-5" title="5">)</a>
<a class="sourceLine" id="cb16-6" title="6">row.min.coef.var &lt;-<span class="st"> </span><span class="kw">which.min</span>(stats.table<span class="op">$</span>noise.threshold.coefficient.of.variation)</a>
<a class="sourceLine" id="cb16-7" title="7"><span class="co"># adjust column names for printing</span></a>
<a class="sourceLine" id="cb16-8" title="8"><span class="kw">colnames</span>(stats.table) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;approach&quot;</span>, <span class="st">&quot;method&quot;</span>, <span class="st">&quot;corr.thr&quot;</span>, <span class="st">&quot;min&quot;</span>, <span class="st">&quot;mean&quot;</span>, <span class="st">&quot;coef.var&quot;</span>, <span class="st">&quot;max&quot;</span>, <span class="st">&quot;all&quot;</span>)</a>
<a class="sourceLine" id="cb16-9" title="9">stats.table[row.min.coef.var, <span class="dv">1</span><span class="op">:</span><span class="dv">7</span>]</a>
<a class="sourceLine" id="cb16-10" title="10">dplyr<span class="op">::</span><span class="kw">filter</span>(stats.table, <span class="kw">round</span>(corr.thr, <span class="dv">2</span>) <span class="op">==</span><span class="st"> </span><span class="fl">0.21</span>)[, <span class="dv">1</span><span class="op">:</span><span class="dv">7</span>]</a>
<a class="sourceLine" id="cb16-11" title="11">dplyr<span class="op">::</span><span class="kw">filter</span>(stats.table, method <span class="op">==</span><span class="st"> &quot;loess10_smoothing&quot;</span>)[, <span class="dv">1</span><span class="op">:</span><span class="dv">7</span>]</a></code></pre></div>
<p>We can then call <em>calculate_noise_threshold_base()</em> with our optimised parameters:</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" title="1">noise.thresholds &lt;-<span class="st"> </span>noisyr<span class="op">::</span><span class="kw">calculate_noise_threshold_base</span>(</a>
<a class="sourceLine" id="cb17-2" title="2">  <span class="dt">expression =</span> expression.summary,</a>
<a class="sourceLine" id="cb17-3" title="3">  <span class="dt">similarity.threshold =</span> <span class="fl">0.21</span>,</a>
<a class="sourceLine" id="cb17-4" title="4">  <span class="dt">method.chosen =</span> <span class="st">&quot;Line_plot-loess10_smoothing&quot;</span></a>
<a class="sourceLine" id="cb17-5" title="5">)</a>
<a class="sourceLine" id="cb17-6" title="6">noise.thresholds</a></code></pre></div>
<p>Noise removal parameter optimisation can be turned on using the <em>minimise.coefficient.of.variation</em> parameter in <em>noisyr()</em> or <em>noisyr_counts()</em> and passing further parameters for <em>similarity.threshold.sequence</em> and/or <em>method.chosen.sequence</em>.</p>
<h2 id="noise-removal">Noise removal</h2>
<p>To produce the denoised count matrix, the function <em>remove_noise_from_matrix()</em> is used with a specified vector of noise thresholds (usually calculated by <em>calculate_noise_threshold_base()</em>).</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb18-1" title="1">expression.matrix.denoised &lt;-<span class="st"> </span>noisyr<span class="op">::</span><span class="kw">remove_noise_from_matrix</span>(</a>
<a class="sourceLine" id="cb18-2" title="2">  <span class="dt">expression.matrix =</span> expression.matrix,</a>
<a class="sourceLine" id="cb18-3" title="3">  <span class="dt">noise.thresholds =</span> noise.thresholds)</a>
<a class="sourceLine" id="cb18-4" title="4"><span class="kw">str</span>(expression.matrix.denoised)</a></code></pre></div>
<p>The behaviour of <em>remove_noise_from_matrix()</em> can be further modified:</p>
<ul>
<li><strong><em>add.threshold</em></strong> whether to add the noise threshold to each entry (default) or set each entry under the noise threshold to the noise threshold.</li>
<li><strong><em>average.threshold</em></strong> whether the noise thresholds for the different samples are averaged (default) or used individually. The latter should be especially avoided if the thresholds have high variance, as it could intoduce artificial differences in the data.<br />
</li>
<li><strong><em>remove.noisy.features</em></strong> whether genes/features whose expression is under the noise threshold in every sample should be removed from the matrix (default) or not</li>
</ul>
<p>Because of these defaults, passing the mean of the thresholds gives a slightly different result (different # of genes fully under the noise threshold:</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" title="1">expression.matrix.denoised.fixed &lt;-<span class="st"> </span>noisyr<span class="op">::</span><span class="kw">remove_noise_from_matrix</span>(</a>
<a class="sourceLine" id="cb19-2" title="2">  <span class="dt">expression.matrix =</span> expression.matrix, </a>
<a class="sourceLine" id="cb19-3" title="3">  <span class="dt">noise.thresholds =</span> <span class="kw">mean</span>(noise.thresholds))</a>
<a class="sourceLine" id="cb19-4" title="4"><span class="kw">nrow</span>(expression.matrix.denoised); <span class="kw">nrow</span>(expression.matrix.denoised.fixed)</a></code></pre></div>
<p>By supplying the corresponding parameters to <em>noisyr()</em>, we obtain the same final result from the full pipeline:</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb20-1" title="1">expression.matrix.denoised.full_pipeline &lt;-<span class="st"> </span>noisyr<span class="op">::</span><span class="kw">noisyr</span>(</a>
<a class="sourceLine" id="cb20-2" title="2">  <span class="dt">approach_for_similarity_calculation =</span> <span class="st">&quot;counts&quot;</span>, </a>
<a class="sourceLine" id="cb20-3" title="3">  <span class="dt">expression.matrix =</span> expression.matrix,</a>
<a class="sourceLine" id="cb20-4" title="4">  <span class="dt">similarity.measure =</span> <span class="st">&quot;correlation_pearson&quot;</span>,</a>
<a class="sourceLine" id="cb20-5" title="5">  <span class="dt">optimise.window.length.logical =</span> <span class="ot">FALSE</span>,</a>
<a class="sourceLine" id="cb20-6" title="6">  <span class="dt">minimise.coefficient.of.variation =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb20-7" title="7">  <span class="dt">similarity.threshold.sequence =</span> <span class="kw">seq</span>(<span class="fl">0.2</span>, <span class="fl">0.3</span>, <span class="dt">by=</span><span class="fl">0.01</span>)</a>
<a class="sourceLine" id="cb20-8" title="8">)</a>
<a class="sourceLine" id="cb20-9" title="9"><span class="kw">identical</span>(expression.matrix.denoised,</a>
<a class="sourceLine" id="cb20-10" title="10">          expression.matrix.denoised.full_pipeline)</a></code></pre></div>
<h1 id="downstream-analysis">Downstream analysis</h1>
<p>The denoised matrix can be used instead of the raw count matrix for downstream analysis. Here we present a simple example of a differential expression (DE) analysis and compare the two.</p>
<p>We create a function to perform the same DE pipeline on both matrices, using <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2796818/">the edgeR package</a>.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb21-1" title="1">DE_edgeR =<span class="st"> </span><span class="cf">function</span>(expression.matrix, metadata){</a>
<a class="sourceLine" id="cb21-2" title="2">  <span class="co"># load or install edgeR</span></a>
<a class="sourceLine" id="cb21-3" title="3">  <span class="cf">if</span>(<span class="op">!</span><span class="kw">require</span>(edgeR)){</a>
<a class="sourceLine" id="cb21-4" title="4">    <span class="cf">if</span> (<span class="op">!</span><span class="kw">requireNamespace</span>(<span class="st">&quot;BiocManager&quot;</span>, <span class="dt">quietly =</span> <span class="ot">TRUE</span>))</a>
<a class="sourceLine" id="cb21-5" title="5">      <span class="kw">install.packages</span>(<span class="st">&quot;BiocManager&quot;</span>)</a>
<a class="sourceLine" id="cb21-6" title="6">    BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="st">&quot;edgeR&quot;</span>)</a>
<a class="sourceLine" id="cb21-7" title="7">  }</a>
<a class="sourceLine" id="cb21-8" title="8">  </a>
<a class="sourceLine" id="cb21-9" title="9">  <span class="co"># create metadata</span></a>
<a class="sourceLine" id="cb21-10" title="10">  metadata &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">id =</span> <span class="kw">colnames</span>(expression.matrix),</a>
<a class="sourceLine" id="cb21-11" title="11">                         <span class="dt">timepoint =</span> <span class="kw">c</span>(<span class="st">&quot;0h&quot;</span>, <span class="st">&quot;0h&quot;</span>, <span class="st">&quot;12h&quot;</span>, <span class="st">&quot;12h&quot;</span>))</a>
<a class="sourceLine" id="cb21-12" title="12">  </a>
<a class="sourceLine" id="cb21-13" title="13">  <span class="co"># quantile normalise</span></a>
<a class="sourceLine" id="cb21-14" title="14">  expression.matrix.normalised &lt;-<span class="st"> </span>preprocessCore<span class="op">::</span><span class="kw">normalize.quantiles</span>(expression.matrix)</a>
<a class="sourceLine" id="cb21-15" title="15">  <span class="kw">rownames</span>(expression.matrix.normalised) &lt;-<span class="st"> </span>base<span class="op">::</span><span class="kw">rownames</span>(expression.matrix)</a>
<a class="sourceLine" id="cb21-16" title="16">  <span class="kw">colnames</span>(expression.matrix.normalised) &lt;-<span class="st"> </span>base<span class="op">::</span><span class="kw">colnames</span>(expression.matrix)</a>
<a class="sourceLine" id="cb21-17" title="17">  </a>
<a class="sourceLine" id="cb21-18" title="18">  <span class="co"># process using edgeR</span></a>
<a class="sourceLine" id="cb21-19" title="19">  expression.matrix.for.de &lt;-<span class="st"> </span><span class="kw">round</span>(expression.matrix.normalised)</a>
<a class="sourceLine" id="cb21-20" title="20">  expression.matrix.for.de &lt;-<span class="st"> </span></a>
<a class="sourceLine" id="cb21-21" title="21"><span class="st">    </span>expression.matrix.for.de[<span class="kw">apply</span>(expression.matrix.for.de, <span class="dv">1</span>, sum) <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>, ]</a>
<a class="sourceLine" id="cb21-22" title="22">  design &lt;-<span class="st"> </span><span class="kw">model.matrix</span>(<span class="op">~</span><span class="st"> </span><span class="dv">0</span> <span class="op">+</span><span class="st"> </span>metadata<span class="op">$</span>timepoint)</a>
<a class="sourceLine" id="cb21-23" title="23">  edger &lt;-<span class="st"> </span><span class="kw">DGEList</span>(<span class="dt">counts =</span> expression.matrix.for.de)</a>
<a class="sourceLine" id="cb21-24" title="24">  edger &lt;-<span class="st"> </span><span class="kw">estimateDisp</span>(edger, design)</a>
<a class="sourceLine" id="cb21-25" title="25">  edger.fit &lt;-<span class="st"> </span><span class="kw">glmFit</span>(edger, design)</a>
<a class="sourceLine" id="cb21-26" title="26">  edger.lrt &lt;-<span class="st"> </span><span class="kw">glmLRT</span>(edger.fit, <span class="dt">contrast=</span><span class="kw">c</span>(<span class="op">-</span><span class="dv">1</span>, <span class="dv">1</span>))</a>
<a class="sourceLine" id="cb21-27" title="27">  </a>
<a class="sourceLine" id="cb21-28" title="28">  <span class="co"># extract results</span></a>
<a class="sourceLine" id="cb21-29" title="29">  res &lt;-<span class="st"> </span><span class="kw">topTags</span>(edger.lrt, <span class="dt">n =</span> <span class="ot">Inf</span>)<span class="op">$</span>table</a>
<a class="sourceLine" id="cb21-30" title="30">  res<span class="op">$</span>DE &lt;-<span class="st"> </span>res<span class="op">$</span>FDR <span class="op">&lt;</span><span class="st"> </span><span class="fl">0.05</span> <span class="op">&amp;</span><span class="st"> </span><span class="kw">abs</span>(res<span class="op">$</span>logFC) <span class="op">&gt;</span><span class="st"> </span><span class="dv">1</span></a>
<a class="sourceLine" id="cb21-31" title="31">  </a>
<a class="sourceLine" id="cb21-32" title="32">  <span class="co"># make volcano plot</span></a>
<a class="sourceLine" id="cb21-33" title="33">  <span class="kw">print</span>(</a>
<a class="sourceLine" id="cb21-34" title="34">    ggplot2<span class="op">::</span><span class="kw">ggplot</span>(res) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb21-35" title="35"><span class="st">      </span>ggplot2<span class="op">::</span><span class="kw">theme_minimal</span>() <span class="op">+</span></a>
<a class="sourceLine" id="cb21-36" title="36"><span class="st">      </span>ggplot2<span class="op">::</span><span class="kw">geom_point</span>(ggplot2<span class="op">::</span><span class="kw">aes</span>(<span class="dt">x=</span>logFC, <span class="dt">y=</span><span class="op">-</span><span class="kw">log10</span>(FDR), <span class="dt">color=</span>DE), <span class="dt">show.legend=</span><span class="ot">FALSE</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb21-37" title="37"><span class="st">      </span>ggplot2<span class="op">::</span><span class="kw">scale_color_manual</span>(<span class="dt">values=</span><span class="kw">c</span>(<span class="st">&quot;black&quot;</span>, <span class="st">&quot;red&quot;</span>)) <span class="op">+</span></a>
<a class="sourceLine" id="cb21-38" title="38"><span class="st">      </span>ggplot2<span class="op">::</span><span class="kw">lims</span>(<span class="dt">x=</span><span class="kw">c</span>(<span class="op">-</span><span class="dv">12</span>, <span class="dv">12</span>), <span class="dt">y=</span><span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">120</span>))</a>
<a class="sourceLine" id="cb21-39" title="39">  )</a>
<a class="sourceLine" id="cb21-40" title="40">  </a>
<a class="sourceLine" id="cb21-41" title="41">  <span class="kw">return</span>(res)</a>
<a class="sourceLine" id="cb21-42" title="42">  </a>
<a class="sourceLine" id="cb21-43" title="43">}</a>
<a class="sourceLine" id="cb21-44" title="44"></a>
<a class="sourceLine" id="cb21-45" title="45">results.raw &lt;-<span class="st"> </span><span class="kw">DE_edgeR</span>(expression.matrix)</a>
<a class="sourceLine" id="cb21-46" title="46">results.denoised &lt;-<span class="st"> </span><span class="kw">DE_edgeR</span>(expression.matrix.denoised)</a></code></pre></div>
<blockquote>
<p>We observe the distribution of genes in the volcano plots becoming a lot tighter for the denoised matrix. For the raw matrix, there are a lot of genes with low p-values and high log-fold changes that are barely called DE. Those “whiskers” are corrected for the denoised matrix.</p>
</blockquote>
<p>We can also see the number of differentially expressed genes has been reduced:</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb22-1" title="1"><span class="kw">sum</span>(results.raw<span class="op">$</span>DE)</a>
<a class="sourceLine" id="cb22-2" title="2"><span class="kw">sum</span>(results.denoised<span class="op">$</span>DE)</a></code></pre></div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
