<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<!-- Inform modern browsers that this page supports both dark and light color schemes,
  and the page author prefers light. --><meta name="color-scheme" content="dark light">
<script>
  // If `prefers-color-scheme` is not supported, fall back to light mode.
  // i.e. In this case, inject the `light` CSS before the others, with
  // no media filter so that it will be downloaded with highest priority.
  if (window.matchMedia("(prefers-color-scheme: dark)").media === "not all") {
    document.documentElement.style.display = "none";
    document.head.insertAdjacentHTML(
      "beforeend",
      "<link id=\"css\" rel=\"stylesheet\" href=\"https://bootswatch.com/3/flatly/bootstrap.css\" onload=\"document.documentElement.style.display = ''\">"
    );
  }
</script><title>noisyR count matrix approach workflow • noisyr</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- Flatly Theme - Light  --><link id="css-light" rel="stylesheet" href="https://bootswatch.com/3/flatly/bootstrap.css" media="(prefers-color-scheme: light), (prefers-color-scheme: no-preference)">
<!-- Darkly Theme - Dark --><link id="css-dark" rel="stylesheet" href="https://bootswatch.com/3/darkly/bootstrap.css" media="(prefers-color-scheme: dark)">
<!-- preferably CSS --><link rel="stylesheet" href="../preferably.css">
<link id="css-code-light" rel="stylesheet" href="../code-color-scheme-light.css" media="(prefers-color-scheme: light), (prefers-color-scheme: no-preference)">
<link id="css-code-dark" rel="stylesheet" href="../code-color-scheme-dark.css" media="(prefers-color-scheme: dark)">
<script src="../darkswitch.js"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="noisyR count matrix approach workflow">
<meta property="og:description" content="noisyr">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">noisyr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/vignette_noisyr_counts.html">noisyR count matrix approach workflow</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
          <a href="#" id="css-toggle-btn">
            <span class="fas fa-adjust fa-lg"></span>
          </a>
        </li>
        
        <li>
  <a href="https://github.com/Core-Bioinformatics/noisyR/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
        
        


      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>noisyR count matrix approach workflow</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/Core-Bioinformatics/noisyR/blob/master/vignettes/vignette_noisyr_counts.Rmd"><code>vignettes/vignette_noisyr_counts.Rmd</code></a></small>
      <div class="hidden name"><code>vignette_noisyr_counts.Rmd</code></div>

    </div>

    
    
<p>The <em>noisyR</em> package is an end-to-end pipeline for quantifying and removing technical noise from HTS datasets. The three main pipeline steps are:</p>
<ol style="list-style-type: decimal">
<li>similarity calculation<br>
</li>
<li>noise quantification<br>
</li>
<li>noise removal</li>
</ol>
<p>Each step can be finely tuned using hyperparameters; optimal, data-driven values for these parameters are also determined.</p>
<p>The package and some applications are described in more detail in <a href="https://www.biorxiv.org/content/10.1101/2021.01.17.427026v2">this paper</a> and is actively maintained on <a href="https://github.com/Core-Bioinformatics/noisyR" class="uri">https://github.com/Core-Bioinformatics/noisyR</a>.</p>
<p>The <strong>count matrix approach</strong> uses the original, un-normalised count matrix, as provided after alignment and feature quantification; each sample is processed individually, only the relative expressions across samples are compared. Relying on the hypothesis that the majority of genes are not DE, most of the evaluations are expected to point towards a high similarity across samples.</p>
<div id="installation" class="section level1">
<h1 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h1>
<p>To install the package, first install all bioconductor dependencies:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">packages.bioc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"preprocessCore"</span>,
                   <span class="st">"IRanges"</span>,
                   <span class="st">"GenomicRanges"</span>,
                   <span class="st">"Rsamtools"</span><span class="op">)</span>
<span class="va">new.packages.bioc</span> <span class="op">&lt;-</span> <span class="va">packages.bioc</span><span class="op">[</span><span class="op">!</span><span class="op">(</span><span class="va">packages.bioc</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html">installed.packages</a></span><span class="op">(</span><span class="op">)</span><span class="op">[</span>,<span class="st">"Package"</span><span class="op">]</span><span class="op">)</span><span class="op">]</span>
<span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">new.packages.bioc</span><span class="op">)</span><span class="op">)</span><span class="op">{</span>
  <span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>
  <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="va">new.packages.bioc</span><span class="op">)</span>
<span class="op">}</span></code></pre></div>
<p>Then, you can install <em>noisyR</em> (and all its other dependencies) from CRAN:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"noisyr"</span><span class="op">)</span></code></pre></div>
<p>To install the latest stable version from GitHub, first install CRAN dependencies:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">packages.cran</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"utils"</span>,
                   <span class="st">"grDevices"</span>,
                   <span class="st">"tibble"</span>,
                   <span class="st">"dplyr"</span>,
                   <span class="st">"magrittr"</span>,
                   <span class="st">"ggplot2"</span>,
                   <span class="st">"philentropy"</span>,
                   <span class="st">"doParallel"</span>,
                   <span class="st">"foreach"</span><span class="op">)</span>
<span class="va">new.packages.cran</span> <span class="op">&lt;-</span> <span class="va">packages.cran</span><span class="op">[</span><span class="op">!</span><span class="op">(</span><span class="va">packages.cran</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html">installed.packages</a></span><span class="op">(</span><span class="op">)</span><span class="op">[</span>,<span class="st">"Package"</span><span class="op">]</span><span class="op">)</span><span class="op">]</span>
<span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">new.packages.cran</span><span class="op">)</span><span class="op">)</span>
  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="va">new.packages.cran</span><span class="op">)</span>

<span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"devtools"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span>
<span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org//reference/remote-reexports.html">install_github</a></span><span class="op">(</span><span class="st">"Core-Bioinformatics/noisyR"</span><span class="op">)</span></code></pre></div>
</div>
<div id="preprocessing" class="section level1">
<h1 class="hasAnchor">
<a href="#preprocessing" class="anchor"></a>Preprocessing</h1>
<p>First, load <em>noisyR</em>:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Core-Bioinformatics/noisyR">noisyr</a></span><span class="op">)</span></code></pre></div>
<p>For this demonstration we will be using a subset of the count matrix for an experiment included in <a href="https://www.sciencedirect.com/science/article/pii/S2405471219301152">a 2019 paper by Yang et al</a>. Rows represent genes/features and columns represent samples:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">counts.in</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"counts_raw.csv"</span>, package <span class="op">=</span> <span class="st">"noisyr"</span><span class="op">)</span>
<span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.csv</a></span><span class="op">(</span><span class="va">counts.in</span>, row.names <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span>
<span class="co">#&gt; 'data.frame':    55573 obs. of  4 variables:</span>
<span class="co">#&gt;  $ SRR7624365: int  2 0 6 0 0 0 0 0 0 0 ...</span>
<span class="co">#&gt;  $ SRR7624366: int  0 0 4 0 0 0 0 0 0 0 ...</span>
<span class="co">#&gt;  $ SRR7624371: int  0 0 2 0 2 0 0 0 0 0 ...</span>
<span class="co">#&gt;  $ SRR7624372: int  0 0 0 0 0 2 0 0 0 0 ...</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span>
<span class="co">#&gt;                    SRR7624365 SRR7624366 SRR7624371 SRR7624372</span>
<span class="co">#&gt; ENSMUSG00000102693          2          0          0          0</span>
<span class="co">#&gt; ENSMUSG00000064842          0          0          0          0</span>
<span class="co">#&gt; ENSMUSG00000051951          6          4          2          0</span>
<span class="co">#&gt; ENSMUSG00000102851          0          0          0          0</span>
<span class="co">#&gt; ENSMUSG00000103377          0          0          2          0</span>
<span class="co">#&gt; ENSMUSG00000104017          0          0          0          2</span></code></pre></div>
<p>Note that when reading from a file R typically returns a data frame. To convert to a matrix we use the function <em>cast_matrix_to_numeric()</em>. This also converts values to numeric (in case they were read as characters). Any values that are not coercible to numeric are replaced by 0.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">expression.matrix</span> <span class="op">&lt;-</span> <span class="fu">noisyr</span><span class="fu">::</span><span class="fu"><a href="../reference/cast_matrix_to_numeric.html">cast_matrix_to_numeric</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span></code></pre></div>
</div>
<div id="running-noisyr" class="section level1">
<h1 class="hasAnchor">
<a href="#running-noisyr" class="anchor"></a>Running <em>noisyR</em>
</h1>
<p>The full <em>noisyR</em> pipeline can be run through <em>noisyr()</em>, choosing “counts” for the count matrix approach (internally calls <em>noisyr_counts()</em>). The user can pass many arguments to this function, which alter the behaviour of the different pipeline steps, as discussed in the breakdown below.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">expression.matrix.denoised.standard</span> <span class="op">&lt;-</span> <span class="fu">noisyr</span><span class="fu">::</span><span class="fu"><a href="../reference/noisyr.html">noisyr</a></span><span class="op">(</span>
  approach_for_similarity_calculation <span class="op">=</span> <span class="st">"counts"</span>, 
  expression.matrix <span class="op">=</span> <span class="va">df</span>
<span class="op">)</span>
<span class="co">#&gt; &gt;&gt;&gt; noisyR counts approach pipeline &lt;&lt;&lt;</span>
<span class="co">#&gt; The input matrix has 55573 rows and 4 cols</span>
<span class="co">#&gt;     number of genes: 55573</span>
<span class="co">#&gt;     number of samples: 4</span>
<span class="co">#&gt; Calculating the number of elements per window</span>
<span class="co">#&gt;     the number of elements per window is 5557</span>
<span class="co">#&gt;     the step size is 277</span>
<span class="co">#&gt;     the selected similarity metric is correlation_pearson</span>
<span class="co">#&gt;   Working with sample 1</span>
<span class="co">#&gt;   Working with sample 2</span>
<span class="co">#&gt;   Working with sample 3</span>
<span class="co">#&gt;   Working with sample 4</span>
<span class="co">#&gt; Calculating noise thresholds for 4 samples...</span>
<span class="co">#&gt;     similarity.threshold = 0.25</span>
<span class="co">#&gt;     method.chosen = Boxplot-IQR</span>
<span class="co">#&gt; Denoising expression matrix...</span>
<span class="co">#&gt;     removing noisy genes</span>
<span class="co">#&gt;     adjusting matrix</span>
<span class="co">#&gt; &gt;&gt;&gt; Done! &lt;&lt;&lt;</span></code></pre></div>
<p>The output of the noise removal is a denoised matrix that can be passed on to other methods for downstream analysis.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">expression.matrix.denoised.standard</span><span class="op">)</span>
<span class="co">#&gt;                    SRR7624365 SRR7624366 SRR7624371 SRR7624372</span>
<span class="co">#&gt; ENSMUSG00000025902         94         74         32         44</span>
<span class="co">#&gt; ENSMUSG00000102269         24         48         20         26</span>
<span class="co">#&gt; ENSMUSG00000098104         76         68         62         60</span>
<span class="co">#&gt; ENSMUSG00000103922        326        273        318        365</span>
<span class="co">#&gt; ENSMUSG00000033845      14451       9925      14702      14549</span>
<span class="co">#&gt; ENSMUSG00000102275        117        100        142        124</span>
<span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="va">expression.matrix.denoised.standard</span>, <span class="fl">2</span>, <span class="va">min</span><span class="op">)</span>
<span class="co">#&gt; SRR7624365 SRR7624366 SRR7624371 SRR7624372 </span>
<span class="co">#&gt;         18         18         18         18</span></code></pre></div>
</div>
<div id="pipeline-breakdown" class="section level1">
<h1 class="hasAnchor">
<a href="#pipeline-breakdown" class="anchor"></a>Pipeline breakdown</h1>
<p>While for most applications, running <em>noisyr()</em> is sufficient, it may be useful to run individual pipeline steps manually. For example, the user may want to create summary figures for the denoising, store the noise thresholds obtained or other intermediary outputs, or try out different options without rerunning all steps. In this section, we look deeper into <em>noisyr()</em> and break down the three main steps that are performed.</p>
<div id="similarity-calculation" class="section level2">
<h2 class="hasAnchor">
<a href="#similarity-calculation" class="anchor"></a>Similarity calculation</h2>
<p>We can then run the similarity calculation using <em>calculate_expression_similarity_counts()</em>:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">expression.summary</span> <span class="op">&lt;-</span> <span class="fu">noisyr</span><span class="fu">::</span><span class="fu"><a href="../reference/calculate_expression_similarity_counts.html">calculate_expression_similarity_counts</a></span><span class="op">(</span>
  expression.matrix <span class="op">=</span> <span class="va">expression.matrix</span>, 
  similarity.measure <span class="op">=</span> <span class="st">"correlation_pearson"</span>
<span class="op">)</span>
<span class="co">#&gt; The input matrix has 55573 rows and 4 cols</span>
<span class="co">#&gt;     number of genes: 55573</span>
<span class="co">#&gt;     number of samples: 4</span>
<span class="co">#&gt; Calculating the number of elements per window</span>
<span class="co">#&gt;     the number of elements per window is 5557</span>
<span class="co">#&gt;     the step size is 277</span>
<span class="co">#&gt;     the selected similarity metric is correlation_pearson</span>
<span class="co">#&gt;   Working with sample 1</span>
<span class="co">#&gt;   Working with sample 2</span>
<span class="co">#&gt;   Working with sample 3</span>
<span class="co">#&gt;   Working with sample 4</span>
<span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">expression.summary</span><span class="op">)</span>
<span class="co">#&gt; List of 3</span>
<span class="co">#&gt;  $ expression.matrix           : num [1:55573, 1:4] 2 0 6 0 0 0 0 0 0 0 ...</span>
<span class="co">#&gt;   ..- attr(*, "dimnames")=List of 2</span>
<span class="co">#&gt;   .. ..$ : chr [1:55573] "ENSMUSG00000102693" "ENSMUSG00000064842" "ENSMUSG00000051951" "ENSMUSG00000102851" ...</span>
<span class="co">#&gt;   .. ..$ : chr [1:4] "SRR7624365" "SRR7624366" "SRR7624371" "SRR7624372"</span>
<span class="co">#&gt;  $ expression.levels           : num [1:181, 1:4] 0 0 0 0 0 0 0 0 0 0 ...</span>
<span class="co">#&gt;  $ expression.levels.similarity: num [1:181, 1:4] NA NA NA NA NA NA NA NA NA NA ...</span></code></pre></div>
<p>Users can select a similarity measure to assess the localised consistency in expression across samples (dissimilarity measures are inverted). See the <em>philentropy</em> package documentation for more information on the different distances. The full list of available metrics can be viewed by:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">noisyr</span><span class="fu">::</span><span class="fu"><a href="../reference/get_methods_correlation_distance.html">get_methods_correlation_distance</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt;  [1] "correlation_pearson"        "correlation_kendall"       </span>
<span class="co">#&gt;  [3] "correlation_spearman"       "distance_euclidean"        </span>
<span class="co">#&gt;  [5] "distance_manhattan"         "distance_minkowski"        </span>
<span class="co">#&gt;  [7] "distance_chebyshev"         "distance_sorensen"         </span>
<span class="co">#&gt;  [9] "distance_gower"             "distance_soergel"          </span>
<span class="co">#&gt; [11] "distance_kulczynski_d"      "distance_canberra"         </span>
<span class="co">#&gt; [13] "distance_lorentzian"        "distance_intersection"     </span>
<span class="co">#&gt; [15] "distance_non-intersection"  "distance_wavehedges"       </span>
<span class="co">#&gt; [17] "distance_czekanowski"       "distance_motyka"           </span>
<span class="co">#&gt; [19] "distance_kulczynski_s"      "distance_tanimoto"         </span>
<span class="co">#&gt; [21] "distance_ruzicka"           "distance_inner_product"    </span>
<span class="co">#&gt; [23] "distance_harmonic_mean"     "distance_cosine"           </span>
<span class="co">#&gt; [25] "distance_hassebrook"        "distance_jaccard"          </span>
<span class="co">#&gt; [27] "distance_dice"              "distance_fidelity"         </span>
<span class="co">#&gt; [29] "distance_bhattacharyya"     "distance_hellinger"        </span>
<span class="co">#&gt; [31] "distance_matusita"          "distance_squared_chord"    </span>
<span class="co">#&gt; [33] "distance_squared_euclidean" "distance_pearson"          </span>
<span class="co">#&gt; [35] "distance_neyman"            "distance_squared_chi"      </span>
<span class="co">#&gt; [37] "distance_prob_symm"         "distance_divergence"       </span>
<span class="co">#&gt; [39] "distance_clark"             "distance_additive_symm"    </span>
<span class="co">#&gt; [41] "distance_kullback-leibler"  "distance_jeffreys"         </span>
<span class="co">#&gt; [43] "distance_k_divergence"      "distance_topsoe"           </span>
<span class="co">#&gt; [45] "distance_jensen-shannon"    "distance_jensen_difference"</span>
<span class="co">#&gt; [47] "distance_taneja"            "distance_kumar-johnson"    </span>
<span class="co">#&gt; [49] "distance_avg"</span></code></pre></div>
<p>By default, the window length is 10% of the number of rows in the matrix, as it has proven effective empirically. A different window can be specified by the <em>n.elements.per.window</em> parameter. The optimal window length can also be estimated by seeking stability of output (but this can be computationally intensive for large datasets):</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">noisyr</span><span class="fu">::</span><span class="fu"><a href="../reference/optimise_window_length.html">optimise_window_length</a></span><span class="op">(</span>
  expression.matrix <span class="op">=</span> <span class="va">expression.matrix</span>,
  similarity.measure <span class="op">=</span> <span class="st">"correlation_pearson"</span>
<span class="op">)</span>
<span class="co">#&gt; Window length optimisation</span>
<span class="co">#&gt;     number of windows: 33</span>
<span class="co">#&gt;     minimum window: 555</span>
<span class="co">#&gt;     maximum window: 18339</span>
<span class="co">#&gt;     window step: 555</span>
<span class="co">#&gt;     # of iterations: 50</span>
<span class="co">#&gt;     minimum similar windows: 3</span>
<span class="co">#&gt; Calculating expression summary and JSE for each window...</span>
<span class="co">#&gt; Performing t-tests...</span>
<span class="co">#&gt; The number of similar windows found for each window (including itself) were:</span>
<span class="co">#&gt;     1 1 1 1 1 2 2 2 2 2 2 3 3 4 2 2 4 4 4 4 4 4 2 5 2 2 2 4 3 4 3 1 2</span>
<span class="co">#&gt; Optimal window found!</span></code></pre></div>
<p><img src="vignette_noisyr_counts_files/figure-html/window_opt-1.png" width="700"></p>
<pre><code>#&gt; [1] 6660</code></pre>
<p>Window length optimisation can be turned on using the <em>optimise.window.length.logical</em> parameter in <em>noisyr()</em> or <em>noisyr_counts()</em>.</p>
<p>Plots of the abundance-correlation relation can be generated through the <em>plot_expression_similarity()</em> function:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">plotlist</span> <span class="op">&lt;-</span> <span class="fu">noisyr</span><span class="fu">::</span><span class="fu"><a href="../reference/plot_expression_similarity.html">plot_expression_similarity</a></span><span class="op">(</span>
  expression.summary <span class="op">=</span> <span class="va">expression.summary</span><span class="op">)</span>
<span class="va">plotlist</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></code></pre></div>
<p><img src="vignette_noisyr_counts_files/figure-html/simple_plot-1.png" width="700"></p>
<blockquote>
<p>As expected, we observe low correlation values for low abundances and a steady increase towards 1 as the abundance increases. This is based on the expectation that most genes are not differentially expressed and have consistent expression, but at low abundances the stochastic nature of transcription and sequencing gives rise to noise. The local maximum at very low abundances is due to strings of zeros driving the correlation higher than expected.</p>
</blockquote>
<p>These are ggplot objects, and can thus be modified and combined intuitively. For example, plotting all the line plots together:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">plotdf.line</span> <span class="op">&lt;-</span> <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span><span class="op">)</span>
<span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">)</span><span class="op">{</span>
  <span class="va">lineid</span> <span class="op">&lt;-</span> <span class="va">i</span> <span class="op">*</span> <span class="fl">2</span> <span class="op">-</span> <span class="fl">1</span>
  <span class="va">plotdf.line</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">rbind</a></span><span class="op">(</span>
    <span class="va">plotdf.line</span>, 
    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span><span class="va">plotlist</span><span class="op">[[</span><span class="va">lineid</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">data</span>,
                  Sample<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">expression.matrix</span><span class="op">)</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>
<span class="op">}</span>

<span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">plotdf.line</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> 
    <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html">geom_line</a></span><span class="op">(</span><span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">x</span>, y<span class="op">=</span><span class="va">y</span>, colour<span class="op">=</span><span class="va">Sample</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_smooth.html">geom_smooth</a></span><span class="op">(</span><span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="va">x</span>,<span class="va">y</span>,colour<span class="op">=</span><span class="va">Sample</span><span class="op">)</span>, method<span class="op">=</span><span class="st">"loess"</span>,
                         formula<span class="op">=</span> <span class="va">y</span> <span class="op">~</span> <span class="va">x</span>, span<span class="op">=</span><span class="fl">0.1</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html">ylim</a></span><span class="op">(</span><span class="fl">0</span><span class="op">:</span><span class="fl">1</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">xlab</a></span><span class="op">(</span><span class="st">"log2(expression)"</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span><span class="op">(</span><span class="st">"Pearson correlation"</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_hline</a></span><span class="op">(</span>yintercept<span class="op">=</span><span class="fl">0.25</span>, color<span class="op">=</span><span class="st">"black"</span><span class="op">)</span></code></pre></div>
<p><img src="vignette_noisyr_counts_files/figure-html/combined_plot-1.png" width="700"></p>
</div>
<div id="noise-quantification" class="section level2">
<h2 class="hasAnchor">
<a href="#noise-quantification" class="anchor"></a>Noise quantification</h2>
<p>Using the output of the similarity calculation, we can compute the signal to noise threshold in each sample:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">noise.thresholds</span> <span class="op">&lt;-</span> <span class="fu">noisyr</span><span class="fu">::</span><span class="fu"><a href="../reference/calculate_noise_threshold_base.html">calculate_noise_threshold_base</a></span><span class="op">(</span>expression <span class="op">=</span> <span class="va">expression.summary</span><span class="op">)</span>
<span class="co">#&gt; Calculating noise thresholds for 4 samples...</span>
<span class="co">#&gt;     similarity.threshold = 0.25</span>
<span class="co">#&gt;     method.chosen = Boxplot-IQR</span>
<span class="va">noise.thresholds</span>
<span class="co">#&gt; [1] 19.69831 25.99208 12.12573 12.99604</span></code></pre></div>
<p>Here we used the default parameters: a similarity threshold of 0.25 and the Boxplot-IQR method. There are several methods available, which can be viewed with the <em>get_methods_calculate_noise_threshold()</em> function:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">noisyr</span><span class="fu">::</span><span class="fu"><a href="../reference/get_methods_calculate_noise_threshold.html">get_methods_calculate_noise_threshold</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt;  [1] "Density_based-No_normalisation"      </span>
<span class="co">#&gt;  [2] "Density_based-RPM_normalisation"     </span>
<span class="co">#&gt;  [3] "Density_based-Quantile_normalisation"</span>
<span class="co">#&gt;  [4] "Line_plot-No_smoothing"              </span>
<span class="co">#&gt;  [5] "Line_plot-loess10_smoothing"         </span>
<span class="co">#&gt;  [6] "Line_plot-loess25_smoothing"         </span>
<span class="co">#&gt;  [7] "Line_plot-loess50_smoothing"         </span>
<span class="co">#&gt;  [8] "Boxplot-Median"                      </span>
<span class="co">#&gt;  [9] "Boxplot-IQR"                         </span>
<span class="co">#&gt; [10] "Boxplot-Quant5"</span></code></pre></div>
<p>The first three methods are just calculating the minimum of the density plot for all genes (a common, fast approach). This usually provides a rough, overestimated signal to noise threshold.</p>
<p>The rest of the methods use either the (smoothed) line plot or the boxplot to find the noise threshold given a similarity (correlation/distance) threshold.</p>
<p>It is recommended that the method with the least coefficient of variation across all samples is chosen for noise removal. This can also be applied to compute the correlation/distance threshold instead of supplying it manually, which is especially useful for non-correlation measures which don’t have a standard range.</p>
<p>For example, by looking to minimise the coefficient of variation, we get a correlation threshold of 0.21 and the loess10 smoothing method for this dataset:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">similarity.threshold.sequence</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="fl">0.2</span>, <span class="fl">0.3</span>, by<span class="op">=</span><span class="fl">0.01</span><span class="op">)</span>
<span class="va">stats.table</span> <span class="op">&lt;-</span> <span class="fu">noisyr</span><span class="fu">::</span><span class="fu"><a href="../reference/calculate_noise_threshold_method_statistics.html">calculate_noise_threshold_method_statistics</a></span><span class="op">(</span>
  expression <span class="op">=</span> <span class="va">expression.summary</span>,
  similarity.threshold.sequence <span class="op">=</span> <span class="va">similarity.threshold.sequence</span>
<span class="op">)</span>
<span class="va">row.min.coef.var</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.min.html">which.min</a></span><span class="op">(</span><span class="va">stats.table</span><span class="op">$</span><span class="va">noise.threshold.coefficient.of.variation</span><span class="op">)</span>
<span class="co"># adjust column names for printing</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">stats.table</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"approach"</span>, <span class="st">"method"</span>, <span class="st">"corr.thr"</span>, <span class="st">"min"</span>, <span class="st">"mean"</span>, <span class="st">"coef.var"</span>, <span class="st">"max"</span>, <span class="st">"all"</span><span class="op">)</span>
<span class="va">stats.table</span><span class="op">[</span><span class="va">row.min.coef.var</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">7</span><span class="op">]</span>
<span class="co">#&gt; # A tibble: 1 x 7</span>
<span class="co">#&gt;   approach  method            corr.thr   min  mean coef.var   max</span>
<span class="co">#&gt;   &lt;chr&gt;     &lt;chr&gt;                &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;    &lt;dbl&gt; &lt;dbl&gt;</span>
<span class="co">#&gt; 1 Line_plot loess10_smoothing     0.21  7.37  9.88    0.253  13.3</span>
<span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">stats.table</span>, <span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="va">corr.thr</span>, <span class="fl">2</span><span class="op">)</span> <span class="op">==</span> <span class="fl">0.21</span><span class="op">)</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">7</span><span class="op">]</span>
<span class="co">#&gt; # A tibble: 10 x 7</span>
<span class="co">#&gt;    approach      method                 corr.thr    min   mean coef.var   max</span>
<span class="co">#&gt;    &lt;chr&gt;         &lt;chr&gt;                     &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;    &lt;dbl&gt; &lt;dbl&gt;</span>
<span class="co">#&gt;  1 Density_based No_normalisation           0.21 308    308      NA     308  </span>
<span class="co">#&gt;  2 Density_based RPM_normalisation          0.21 290    290      NA     290  </span>
<span class="co">#&gt;  3 Density_based Quantile_normalisation     0.21 304    304      NA     304  </span>
<span class="co">#&gt;  4 Line_plot     No_smoothing               0.21   0      0     NaN       0  </span>
<span class="co">#&gt;  5 Line_plot     loess10_smoothing          0.21   7.37   9.88    0.253  13.3</span>
<span class="co">#&gt;  6 Line_plot     loess25_smoothing          0.21   4.26   9.20    0.450  14.3</span>
<span class="co">#&gt;  7 Line_plot     loess50_smoothing          0.21   6.47   9.92    0.336  14.3</span>
<span class="co">#&gt;  8 Boxplot       Median                     0.21   8.57  11.6     0.271  16  </span>
<span class="co">#&gt;  9 Boxplot       IQR                        0.21   8.57  11.6     0.271  16  </span>
<span class="co">#&gt; 10 Boxplot       Quant5                     0.21   8.57  11.6     0.271  16</span>
<span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">stats.table</span>, <span class="va">method</span> <span class="op">==</span> <span class="st">"loess10_smoothing"</span><span class="op">)</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">7</span><span class="op">]</span>
<span class="co">#&gt; # A tibble: 11 x 7</span>
<span class="co">#&gt;    approach  method            corr.thr   min  mean coef.var   max</span>
<span class="co">#&gt;    &lt;chr&gt;     &lt;chr&gt;                &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;    &lt;dbl&gt; &lt;dbl&gt;</span>
<span class="co">#&gt;  1 Line_plot loess10_smoothing     0.2   3.72  8.42    0.425  12.4</span>
<span class="co">#&gt;  2 Line_plot loess10_smoothing     0.21  7.37  9.88    0.253  13.3</span>
<span class="co">#&gt;  3 Line_plot loess10_smoothing     0.22  8.42 11.3     0.321  16.6</span>
<span class="co">#&gt;  4 Line_plot loess10_smoothing     0.23  9.02 12.7     0.358  19.3</span>
<span class="co">#&gt;  5 Line_plot loess10_smoothing     0.24 10.4  14.3     0.395  22.5</span>
<span class="co">#&gt;  6 Line_plot loess10_smoothing     0.25 10.6  16.1     0.388  24.3</span>
<span class="co">#&gt;  7 Line_plot loess10_smoothing     0.26 11.4  17.9     0.433  28.4</span>
<span class="co">#&gt;  8 Line_plot loess10_smoothing     0.27 11.4  19.4     0.439  30.8</span>
<span class="co">#&gt;  9 Line_plot loess10_smoothing     0.28 12.2  24.0     0.578  43.3</span>
<span class="co">#&gt; 10 Line_plot loess10_smoothing     0.29 15.9  28.0     0.505  47.4</span>
<span class="co">#&gt; 11 Line_plot loess10_smoothing     0.3  17.1  35.4     0.486  52.0</span></code></pre></div>
<p>We can then call <em>calculate_noise_threshold_base()</em> with our optimised parameters:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">noise.thresholds</span> <span class="op">&lt;-</span> <span class="fu">noisyr</span><span class="fu">::</span><span class="fu"><a href="../reference/calculate_noise_threshold_base.html">calculate_noise_threshold_base</a></span><span class="op">(</span>
  expression <span class="op">=</span> <span class="va">expression.summary</span>,
  similarity.threshold <span class="op">=</span> <span class="fl">0.21</span>,
  method.chosen <span class="op">=</span> <span class="st">"Line_plot-loess10_smoothing"</span>
<span class="op">)</span>
<span class="co">#&gt; Calculating noise thresholds for 4 samples...</span>
<span class="co">#&gt;     similarity.threshold = 0.21</span>
<span class="co">#&gt;     method.chosen = Line_plot-loess10_smoothing</span>
<span class="va">noise.thresholds</span>
<span class="co">#&gt; [1]  9.664027 13.325175  7.368364  9.171495</span></code></pre></div>
<p>Noise removal parameter optimisation can be turned on using the <em>minimise.coefficient.of.variation</em> parameter in <em>noisyr()</em> or <em>noisyr_counts()</em> and passing further parameters for <em>similarity.threshold.sequence</em> and/or <em>method.chosen.sequence</em>.</p>
</div>
<div id="noise-removal" class="section level2">
<h2 class="hasAnchor">
<a href="#noise-removal" class="anchor"></a>Noise removal</h2>
<p>To produce the denoised count matrix, the function <em>remove_noise_from_matrix()</em> is used with a specified vector of noise thresholds (usually calculated by <em>calculate_noise_threshold_base()</em>).</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">expression.matrix.denoised</span> <span class="op">&lt;-</span> <span class="fu">noisyr</span><span class="fu">::</span><span class="fu"><a href="../reference/remove_noise_from_matrix.html">remove_noise_from_matrix</a></span><span class="op">(</span>
  expression.matrix <span class="op">=</span> <span class="va">expression.matrix</span>,
  noise.thresholds <span class="op">=</span> <span class="va">noise.thresholds</span><span class="op">)</span>
<span class="co">#&gt; Denoising expression matrix...</span>
<span class="co">#&gt;     removing noisy genes</span>
<span class="co">#&gt;     adjusting matrix</span>
<span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">expression.matrix.denoised</span><span class="op">)</span>
<span class="co">#&gt;  num [1:24341, 1:4] 17 86 16 68 318 ...</span>
<span class="co">#&gt;  - attr(*, "dimnames")=List of 2</span>
<span class="co">#&gt;   ..$ : chr [1:24341] "ENSMUSG00000102343" "ENSMUSG00000025902" "ENSMUSG00000102269" "ENSMUSG00000098104" ...</span>
<span class="co">#&gt;   ..$ : chr [1:4] "SRR7624365" "SRR7624366" "SRR7624371" "SRR7624372"</span></code></pre></div>
<p>The behaviour of <em>remove_noise_from_matrix()</em> can be further modified:</p>
<ul>
<li>
<strong><em>add.threshold</em></strong> whether to add the noise threshold to each entry (default) or set each entry under the noise threshold to the noise threshold.</li>
<li>
<strong><em>average.threshold</em></strong> whether the noise thresholds for the different samples are averaged (default) or used individually. The latter should be especially avoided if the thresholds have high variance, as it could intoduce artificial differences in the data.<br>
</li>
<li>
<strong><em>remove.noisy.features</em></strong> whether genes/features whose expression is under the noise threshold in every sample should be removed from the matrix (default) or not</li>
</ul>
<p>Because of these defaults, passing the mean of the thresholds gives a slightly different result (different # of genes fully under the noise threshold:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">expression.matrix.denoised.fixed</span> <span class="op">&lt;-</span> <span class="fu">noisyr</span><span class="fu">::</span><span class="fu"><a href="../reference/remove_noise_from_matrix.html">remove_noise_from_matrix</a></span><span class="op">(</span>
  expression.matrix <span class="op">=</span> <span class="va">expression.matrix</span>, 
  noise.thresholds <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">noise.thresholds</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; noise.thresholds only has 1 value, using a fixed threshold...</span>
<span class="co">#&gt; Denoising expression matrix...</span>
<span class="co">#&gt;     removing noisy genes</span>
<span class="co">#&gt;     adjusting matrix</span>
<span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">expression.matrix.denoised</span><span class="op">)</span>; <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">expression.matrix.denoised.fixed</span><span class="op">)</span>
<span class="co">#&gt; [1] 24341</span>
<span class="co">#&gt; [1] 24309</span></code></pre></div>
<p>By supplying the corresponding parameters to <em>noisyr()</em>, we obtain the same final result from the full pipeline:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">expression.matrix.denoised.full_pipeline</span> <span class="op">&lt;-</span> <span class="fu">noisyr</span><span class="fu">::</span><span class="fu"><a href="../reference/noisyr.html">noisyr</a></span><span class="op">(</span>
  approach_for_similarity_calculation <span class="op">=</span> <span class="st">"counts"</span>, 
  expression.matrix <span class="op">=</span> <span class="va">expression.matrix</span>,
  similarity.measure <span class="op">=</span> <span class="st">"correlation_pearson"</span>,
  optimise.window.length.logical <span class="op">=</span> <span class="cn">FALSE</span>,
  minimise.coefficient.of.variation <span class="op">=</span> <span class="cn">TRUE</span>,
  similarity.threshold.sequence <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="fl">0.2</span>, <span class="fl">0.3</span>, by<span class="op">=</span><span class="fl">0.01</span><span class="op">)</span>
<span class="op">)</span>
<span class="co">#&gt; &gt;&gt;&gt; noisyR counts approach pipeline &lt;&lt;&lt;</span>
<span class="co">#&gt; The input matrix has 55573 rows and 4 cols</span>
<span class="co">#&gt;     number of genes: 55573</span>
<span class="co">#&gt;     number of samples: 4</span>
<span class="co">#&gt; Calculating the number of elements per window</span>
<span class="co">#&gt;     the number of elements per window is 5557</span>
<span class="co">#&gt;     the step size is 277</span>
<span class="co">#&gt;     the selected similarity metric is correlation_pearson</span>
<span class="co">#&gt;   Working with sample 1</span>
<span class="co">#&gt;   Working with sample 2</span>
<span class="co">#&gt;   Working with sample 3</span>
<span class="co">#&gt;   Working with sample 4</span>
<span class="co">#&gt; Selecting parameters that minimise the coefficient of variation...</span>
<span class="co">#&gt; Calculating noise thresholds for 4 samples...</span>
<span class="co">#&gt;     similarity.threshold = 0.21</span>
<span class="co">#&gt;     method.chosen = Line_plot-loess10_smoothing</span>
<span class="co">#&gt; Denoising expression matrix...</span>
<span class="co">#&gt;     removing noisy genes</span>
<span class="co">#&gt;     adjusting matrix</span>
<span class="co">#&gt; &gt;&gt;&gt; Done! &lt;&lt;&lt;</span>
<span class="fu"><a href="https://rdrr.io/r/base/identical.html">identical</a></span><span class="op">(</span><span class="va">expression.matrix.denoised</span>,
          <span class="va">expression.matrix.denoised.full_pipeline</span><span class="op">)</span>
<span class="co">#&gt; [1] TRUE</span></code></pre></div>
</div>
</div>
<div id="downstream-analysis" class="section level1">
<h1 class="hasAnchor">
<a href="#downstream-analysis" class="anchor"></a>Downstream analysis</h1>
<p>The denoised matrix can be used instead of the raw count matrix for downstream analysis. Here we present a simple example of a differential expression (DE) analysis and compare the two.</p>
<p>We create a function to perform the same DE pipeline on both matrices, using <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2796818/">the edgeR package</a>.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">DE_edgeR</span> <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">expression.matrix</span>, <span class="va">metadata</span><span class="op">)</span><span class="op">{</span>
  <span class="co"># load or install edgeR</span>
  <span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span><span class="op">(</span><span class="va"><a href="http://bioinf.wehi.edu.au/edgeR">edgeR</a></span><span class="op">)</span><span class="op">)</span><span class="op">{</span>
    <span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
      <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>
    <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="st">"edgeR"</span><span class="op">)</span>
  <span class="op">}</span>
  
  <span class="co"># create metadata</span>
  <span class="va">metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">expression.matrix</span><span class="op">)</span>,
                         timepoint <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"0h"</span>, <span class="st">"0h"</span>, <span class="st">"12h"</span>, <span class="st">"12h"</span><span class="op">)</span><span class="op">)</span>
  
  <span class="co"># quantile normalise</span>
  <span class="va">expression.matrix.normalised</span> <span class="op">&lt;-</span> <span class="fu">preprocessCore</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/preprocessCore/man/normalize.quantiles.html">normalize.quantiles</a></span><span class="op">(</span><span class="va">expression.matrix</span><span class="op">)</span>
  <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">expression.matrix.normalised</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">base</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">expression.matrix</span><span class="op">)</span>
  <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">expression.matrix.normalised</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">base</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">expression.matrix</span><span class="op">)</span>
  
  <span class="co"># process using edgeR</span>
  <span class="va">expression.matrix.for.de</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="va">expression.matrix.normalised</span><span class="op">)</span>
  <span class="va">expression.matrix.for.de</span> <span class="op">&lt;-</span> 
    <span class="va">expression.matrix.for.de</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="va">expression.matrix.for.de</span>, <span class="fl">1</span>, <span class="va">sum</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span>, <span class="op">]</span>
  <span class="va">design</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html">model.matrix</a></span><span class="op">(</span><span class="op">~</span> <span class="fl">0</span> <span class="op">+</span> <span class="va">metadata</span><span class="op">$</span><span class="va">timepoint</span><span class="op">)</span>
  <span class="va">edger</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/DGEList.html">DGEList</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">expression.matrix.for.de</span><span class="op">)</span>
  <span class="va">edger</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/estimateDisp.html">estimateDisp</a></span><span class="op">(</span><span class="va">edger</span>, <span class="va">design</span><span class="op">)</span>
  <span class="va">edger.fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/glmfit.html">glmFit</a></span><span class="op">(</span><span class="va">edger</span>, <span class="va">design</span><span class="op">)</span>
  <span class="va">edger.lrt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/glmfit.html">glmLRT</a></span><span class="op">(</span><span class="va">edger.fit</span>, contrast<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span>
  
  <span class="co"># extract results</span>
  <span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/topTags.html">topTags</a></span><span class="op">(</span><span class="va">edger.lrt</span>, n <span class="op">=</span> <span class="cn">Inf</span><span class="op">)</span><span class="op">$</span><span class="va">table</span>
  <span class="va">res</span><span class="op">$</span><span class="va">DE</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">FDR</span> <span class="op">&lt;</span> <span class="fl">0.05</span> <span class="op">&amp;</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">logFC</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">1</span>
  
  <span class="co"># make volcano plot</span>
  <span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span>
    <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span> <span class="op">+</span> 
      <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
      <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">logFC</span>, y<span class="op">=</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">FDR</span><span class="op">)</span>, color<span class="op">=</span><span class="va">DE</span><span class="op">)</span>, show.legend<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span>
      <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span><span class="op">(</span>values<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"red"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
      <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html">lims</a></span><span class="op">(</span>x<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">12</span>, <span class="fl">12</span><span class="op">)</span>, y<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">120</span><span class="op">)</span><span class="op">)</span>
  <span class="op">)</span>
  
  <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span>
  
<span class="op">}</span>

<span class="va">results.raw</span> <span class="op">&lt;-</span> <span class="fu">DE_edgeR</span><span class="op">(</span><span class="va">expression.matrix</span><span class="op">)</span></code></pre></div>
<p><img src="vignette_noisyr_counts_files/figure-html/edgeR-1.png" width="700"></p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">results.denoised</span> <span class="op">&lt;-</span> <span class="fu">DE_edgeR</span><span class="op">(</span><span class="va">expression.matrix.denoised</span><span class="op">)</span></code></pre></div>
<p><img src="vignette_noisyr_counts_files/figure-html/edgeR-2.png" width="700"></p>
<blockquote>
<p>We observe the distribution of genes in the volcano plots becoming a lot tighter for the denoised matrix. For the raw matrix, there are a lot of genes with low p-values and high log-fold changes that are barely called DE. Those “whiskers” are corrected for the denoised matrix.</p>
</blockquote>
<p>We can also see the number of differentially expressed genes has been reduced:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">results.raw</span><span class="op">$</span><span class="va">DE</span><span class="op">)</span>
<span class="co">#&gt; [1] 3810</span>
<span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">results.denoised</span><span class="op">$</span><span class="va">DE</span><span class="op">)</span>
<span class="co">#&gt; [1] 3347</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Ilias Moutsopoulos, Irina Mohorianu.</p>
</div>

<div class="pkgdown">
  <p>Made with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1, using <a href="https://preferably.amirmasoudabdol.name/?source=footer">preferably</a> template.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
