<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<!-- Inform modern browsers that this page supports both dark and light color schemes,
  and the page author prefers light. -->
<meta name="color-scheme" content="dark light">
<script>
  // If `prefers-color-scheme` is not supported, fall back to light mode.
  // i.e. In this case, inject the `light` CSS before the others, with
  // no media filter so that it will be downloaded with highest priority.
  if (window.matchMedia("(prefers-color-scheme: dark)").media === "not all") {
    document.documentElement.style.display = "none";
    document.head.insertAdjacentHTML(
      "beforeend",
      "<link id=\"css\" rel=\"stylesheet\" href=\"https://bootswatch.com/3/flatly/bootstrap.css\" onload=\"document.documentElement.style.display = ''\">"
    );
  }
</script>

<title>Calcualte the distance matrices using the BAM files — calculate_expression_similarity_transcript • noisyr</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- Flatly Theme - Light  -->
<link id="css-light" rel="stylesheet" href="https://bootswatch.com/3/flatly/bootstrap.css" media="(prefers-color-scheme: light), (prefers-color-scheme: no-preference)">

<!-- Darkly Theme - Dark -->
<link id="css-dark" rel="stylesheet" href="https://bootswatch.com/3/darkly/bootstrap.css" media="(prefers-color-scheme: dark)">

<!-- preferably CSS -->
<link rel="stylesheet" href="../preferably.css">
<link id="css-code-light" rel="stylesheet" href="../code-color-scheme-light.css" media="(prefers-color-scheme: light), (prefers-color-scheme: no-preference)">
<link id="css-code-dark" rel="stylesheet" href="../code-color-scheme-dark.css" media="(prefers-color-scheme: dark)">

  <script src="../darkswitch.js"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Calcualte the distance matrices using the BAM files — calculate_expression_similarity_transcript" />
<meta property="og:description" content="This function generates an average correlation/distance coefficient
for every exon present in the BAM files. This is done by calculating
the point-to-point correlation/distance of the distribution of reads
across the transcript of each exon and comparing it across samples.
The reason why exons are used instead of full length genes is that long
intronic regions artificially increase the correlation since there is
consistently no expression there, across samples. The user has the
option to use genes instead, by running cast_gtf_to_genes() separately,
with non default parameters." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>




<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">noisyr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/vignette_noisyr_counts.html">noisyR count matrix approach workflow</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
        <li>
          <a href="#" id="css-toggle-btn">
            <span class="fas fa-adjust fa-lg"></span>
          </a>
        </li>
        
        <li>
  <a href="https://github.com/Core-Bioinformatics/noisyR/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
        
        


      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Calcualte the distance matrices using the BAM files</h1>
    <small class="dont-index">Source: <a href='https://github.com/Core-Bioinformatics/noisyR/blob/master/R/calculate_expression_similarity_transcript.R'><code>R/calculate_expression_similarity_transcript.R</code></a></small>
    <div class="hidden name"><code>calculate_expression_similarity_transcript.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function generates an average correlation/distance coefficient
for every exon present in the BAM files. This is done by calculating
the point-to-point correlation/distance of the distribution of reads
across the transcript of each exon and comparing it across samples.
The reason why exons are used instead of full length genes is that long
intronic regions artificially increase the correlation since there is
consistently no expression there, across samples. The user has the
option to use genes instead, by running cast_gtf_to_genes() separately,
with non default parameters.</p>
    </div>

    <pre class="usage"><span class='fu'>calculate_expression_similarity_transcript</span><span class='op'>(</span>
  bams <span class='op'>=</span> <span class='cn'>NULL</span>,
  path.bams <span class='op'>=</span> <span class='st'>"."</span>,
  genes <span class='op'>=</span> <span class='cn'>NULL</span>,
  path.gtf <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.files.html'>list.files</a></span><span class='op'>(</span><span class='st'>"."</span>, pattern <span class='op'>=</span> <span class='st'>"\\.g[tf]f$"</span><span class='op'>)</span>,
  expression.matrix <span class='op'>=</span> <span class='cn'>NULL</span>,
  subsample.genes <span class='op'>=</span> <span class='cn'>FALSE</span>,
  make.index <span class='op'>=</span> <span class='cn'>FALSE</span>,
  unique.only <span class='op'>=</span> <span class='cn'>TRUE</span>,
  mapq.unique <span class='op'>=</span> <span class='fl'>255</span>,
  slack <span class='op'>=</span> <span class='fl'>200</span>,
  similarity.measure <span class='op'>=</span> <span class='st'>"correlation_pearson"</span>,
  save.image.every.1000 <span class='op'>=</span> <span class='cn'>FALSE</span>,
  ncores <span class='op'>=</span> <span class='fl'>1</span>,
  <span class='va'>...</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>bams, path.bams</th>
      <td><p>either a path to the directory where the BAM files are
or a vector of paths to each individual file; if a path is specified,
it extracts all files that end in .bam; looks in the working directory by default</p></td>
    </tr>
    <tr>
      <th>genes</th>
      <td><p>a tibble of the exons extracted from the gtf file;
this is meant for speed if the output of cast_gtf_to_genes() is already generated,
or if the user wants to only calculate similarity for a subset of exons</p></td>
    </tr>
    <tr>
      <th>path.gtf</th>
      <td><p>the path to the gtf/gff annotation file (only used if genes is not
provided); if unspecified, looks for one in the working directory</p></td>
    </tr>
    <tr>
      <th>expression.matrix</th>
      <td><p>expression matrix; not necessary but is used to filter the
gtf to fewer entries and for subsampling if subsample.genes=TRUE;
if not provided, raw read counts per exon are extracted from the BAM files</p></td>
    </tr>
    <tr>
      <th>subsample.genes</th>
      <td><p>logical, whether to subsample low abundance genes to decrease
computational time; the first minimum of the distribution of abundances is calculated,
and genes lower than it are subsampled to match the number of genes higher than it;
the expression matrix needs to be provided for this calculation;
a plot is generated to show that minimum</p></td>
    </tr>
    <tr>
      <th>make.index</th>
      <td><p>whether a BAM index should be generated; if this is FALSE (the default)
and no index exists, the function will exit with an error; the index needs to have
the same name as each BAM file, but ending with .bam.bai</p></td>
    </tr>
    <tr>
      <th>unique.only</th>
      <td><p>whether only uniquely mapped reads should contribute to the expression
of a gene/exon; default is TRUE</p></td>
    </tr>
    <tr>
      <th>mapq.unique</th>
      <td><p>The values of the mapping quality field in the BAM file that corresponds
to uniquely mapped reads; by default, values of 255 are used as these correspond to
the most popular aligners, but an adjustment might be needed;
the mapq scores should be as follows: 255 for STAR, 60 for hisat2,
255 for bowtie in -k mode, 40 for bowtie2 default, 50 for tophat</p></td>
    </tr>
    <tr>
      <th>slack</th>
      <td><p>slack needs to be &gt;=readLength, adjust for efficiency; the default is 200,
as it is higher than most modern sequencing experiments</p></td>
    </tr>
    <tr>
      <th>similarity.measure</th>
      <td><p>one of the similarity metrics to be used, defaults to pearson correlation;
currently, only correlation is supported</p></td>
    </tr>
    <tr>
      <th>save.image.every.1000</th>
      <td><p>whether to save a workspace image after every 1000 exons
are processed; default is FALSE</p></td>
    </tr>
    <tr>
      <th>ncores</th>
      <td><p>Number of cores for parallel computation; defaults to sequential computation,
but parallelisation is highly encouraged; it is set to detectCores() if higher</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>arguments passed on to other methods</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A list with three elements: the first element is the expression matrix,
        as supplied or calculated; the other two are the expression levels matrix and
        expression levels similarity matrix;
        they have the same # of columns as the expression matrix,
        and as many rows as exons processed.</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='va'>bams</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/rep.html'>rep</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>"extdata"</span>, <span class='st'>"ex1.bam"</span>, package<span class='op'>=</span><span class='st'>"Rsamtools"</span>, mustWork<span class='op'>=</span><span class='cn'>TRUE</span><span class='op'>)</span>, <span class='fl'>2</span><span class='op'>)</span>
<span class='va'>genes</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/data.frame.html'>data.frame</a></span><span class='op'>(</span><span class='st'>"id"</span> <span class='op'>=</span> <span class='fl'>1</span><span class='op'>:</span><span class='fl'>2</span>,
                    <span class='st'>"gene_id"</span> <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"gene1"</span>, <span class='st'>"gene2"</span><span class='op'>)</span>,
                    <span class='st'>"seqid"</span> <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"seq1"</span>, <span class='st'>"seq2"</span><span class='op'>)</span>,
                    <span class='st'>"start"</span> <span class='op'>=</span> <span class='fl'>1</span>,
                    <span class='st'>"end"</span> <span class='op'>=</span> <span class='fl'>1600</span><span class='op'>)</span>
<span class='va'>expression.summary</span> <span class='op'>&lt;-</span> <span class='fu'>calculate_expression_similarity_transcript</span><span class='op'>(</span>
  bams <span class='op'>=</span> <span class='va'>bams</span>,
  genes <span class='op'>=</span> <span class='va'>genes</span>,
  mapq.unique <span class='op'>=</span> <span class='fl'>99</span>
<span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>Calculating expression similarity for 2 genes...</span></div><div class='output co'>#&gt; <span class='message'>    this process may take a long time...</span></div><div class='output co'>#&gt; <span class='message'>    ncores=1, running sequentially...</span></div><div class='output co'>#&gt; <span class='message'>Finished! Time elapsed: 0.54 secs</span></div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Ilias Moutsopoulos, Irina Mohorianu.</p>
</div>

<div class="pkgdown">
  <p>Made with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1, using <a href="https://preferably.amirmasoudabdol.name/?source=footer">preferably</a> template.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


